ocaml/xapi/repository_helpers.ml

```diff
@@ -684,27 +698,56 @@ let get_cachedir repo_name =
         let msg = Printf.sprintf "Not found cachedir for repository %s" repo_name in
         raise Api_errors.(Server_error (internal_error, [msg])))
 
-let with_local_repository ~__context f =
+let get_enabled_repositories ~__context =
+  let pool = Helpers.get_pool ~__context in
+  Db.Pool.get_repositories ~__context ~self:pool
+
+let get_remote_repository_name ~__context ~self =
+    !Xapi_globs.remote_repository_prefix ^ "-" ^ (Db.Repository.get_name_label ~__context ~self)
+
+let get_local_repository_name ~__context ~self =
+    !Xapi_globs.local_repository_prefix ^ "-" ^ (Db.Repository.get_name_label ~__context ~self)
+
+let with_local_repositories ~__context f =
   let master_addr =
     try Pool_role.get_master_address ()
     with Pool_role.This_host_is_a_master ->
       Option.get (Helpers.get_management_ip_addr ~__context)
   in
-  let url = Printf.sprintf "https://%s/repository/" master_addr in
+  let enabled = get_enabled_repositories ~__context in
   Xapi_stdext_pervasives.Pervasiveext.finally
     (fun () ->
-      remove_repo_conf_file !Xapi_globs.local_repo_name;
-      write_yum_config url !Xapi_globs.local_repo_name;
-      let config_params =
-          [
-            "--save";
-            Printf.sprintf "--setopt=%s.sslverify=false" !Xapi_globs.local_repo_name;
-            !Xapi_globs.local_repo_name
-          ]
-      in
-      ignore (Helpers.call_script !Xapi_globs.yum_config_manager_cmd config_params);
-      f ())
-    (fun () -> remove_repo_conf_file !Xapi_globs.local_repo_name)
+       let repositories =
+         List.map
+           (fun repository ->
+              let repo_name = get_local_repository_name ~__context ~self:repository in
+              let url =
+                Printf.sprintf "https://%s/repository/%s/"
+                  master_addr
+                  (get_remote_repository_name ~__context ~self:repository)
+              in
+              remove_repo_conf_file repo_name ;
+              write_yum_config url repo_name ;
+              clean_yum_cache repo_name ;
+              let config_params =
+                [
+                  "--save"
+                ; Printf.sprintf "--setopt=%s.sslverify=false" repo_name
+                ; repo_name
+                ]
+              in
+              ignore (Helpers.call_script !Xapi_globs.yum_config_manager_cmd config_params) ;
```

Comment:
If this fails with an exception, what happens?

Reply:
The exception will be captured by difference API functions and raise proper error based on the specific API function.
In API function handler get_host_updates_in_json, it will raise get_host_updates_failed.
In API function handler Repository.apply_updates, it will raise apply_updates_failed
And the error message will be logged.